### PR TITLE
Refactored Polycos

### LINKED ISSUE #847: Bugs in Polyco writer
Here's a comparison between Polyco entries generated with the same parameters between `tempo2` and `PINT`:

```
# tempo2 -tempo1 -f test.par -polyco ...

B1937+21    5-May-18  223000.00   58243.93750000000            71.022344 -0.715 -6.258
 146695203280.882825  641.928232294317   ao   72   12   327.000                
  7.58475477599393022e-07  2.75691843068349529e+00  1.05182499406509443e-04
 -5.53800809293190177e-08 -1.79274838947070341e-10 -1.40505010500570550e-12
  1.44770018627411138e-14  2.02242076264123854e-15 -1.29392880277010108e-17
 -1.27054996558261958e-18  3.76553463657722022e-21  2.91010567554812097e-22

# pint.polycos.Polycos.write_polyco_file()

B1937+21  06-May-18 003600.00 58244.025 71.02227638          0.0   0.0
146700056607.221987 641.928232294317  ao   7      12   327.0                0.0  
4.21842087313951317e-08  2.77946124184379206e+00  6.89972511753701793e-05  
-1.30719833893706870e-07 -1.21324048857498680e-10 1.49527601758533278e-14  
1.17225403493628147e-14  1.20414395241383120e-16  -1.05153148084523127e-17 
-3.95887144140585074e-20 3.32837169386443349e-21  -6.16403723498682311e-25 
```

PINT's Polyco file writer fail to match the [format specification](http://tempo.sourceforge.net/ref_man_sections/tz-polyco.txt) in terms of positions and lengths of segments.

But an even more serious bug: The `nspan` value which should be `72` gets truncated to `7` in PINT's output!

I would be happy to submit a pull request to fix these issues. If there are additional thoughts/suggestions on these bugs, I'd be happy to incorporate them in my PR.

PS: Funnily, `tempo2`'s output doesn't exact follow the format specification either (and the specs seem to be incomplete as well when it comes to binary orbit parameters)! I would love to know if there exists a better, more updated specification for the Polyco format.  Weirdly, even the comments on `polyco.C` in the current `tempo2` repository on bitbucket are also incorrect!

### PR DESCRIPTION
**Pull request by theXYZT**
Resolves #847.

I refactored `polycos.py` to match the behavior of tempo2. Since the [polyco format specification](http://tempo.sourceforge.net/ref_man_sections/tz-polyco.txt) is incorrect, I mimicked [the implementation in polyco.C in the official tempo2 repository](https://bitbucket.org/psrsoft/tempo2/src/master/polyco.C). 

In some unlikely edge cases, the tempo2 implementation has some bugs that break things - those aren't replicated here.

I added a couple of simple round-trip tests and some other basic tests for coverage.

### CODE REVIEW

**Review by aarchiba — COMMENTED** [review]
It looks good! I like the round-trip testing.

File: src/pint/polycos.py [hunk 1]
```diff
@@ -612,88 +571,69 @@ def generate_polycos(
                 toas = toa.get_TOAs_list(toaList)
 
                 ph = model.phase(toas)
-                dt = (nodes * u.day - tmid).to("min")  # Use constant
+                dt = (nodes - tmid) * 1440.0
```
[the remark this answers is not in the document]
Comment (theXYZT) [reply src/pint/polycos.py:574]: I can set a variable in the module namespace called "MIN_PER_DAY" and set it to that. Will do.

File: src/pint/polycos.py [hunk 2]
```diff
@@ -612,88 +571,69 @@ def generate_polycos(
...
-                )
-                date, hms = midTime.iso.split()
+                coeffs = np.polyfit(dtd, rdcPhased, ncoeff - 1)[::-1]
+
+                date, hms = Time(tmid, format="mjd", scale="utc").iso.split()
                 yy, mm, dd = date.split("-")
-                date = dd + "-" + month[int(mm) - 1] + "-" + yy[2:4]
-                hms = hms.replace(":", "")
+                date = dd + "-" + MONTHS[int(mm) - 1] + "-" + yy[-2:]
+                hms = data2longdouble(hms.replace(":", ""))
```
[the remark this answers is not in the document]
Comment (theXYZT) [reply src/pint/polycos.py:584]: UTC is used for nothing but writing and reading from a polyco entry. This is how tempo's polyco format works - it actually wants to print a float. 

While working on this PR, I was constantly confused at the many weird design decisions of the polyco format.

PS: It doesn't really need to be a longdouble though - I just did that for consistency with how the polyco module was coded before (a lot of things were unnecessarily cast to longdouble). I can downgrade it to a simple float.

File: tests/test_polycos.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    model = get_model(str(par_file))
+
+    p = Polycos()
+    p.generate_polycos(model, 55000, 55002, "ao", 300, 12, 1400.0)
+    p.write_polyco_file(output_polyco)
+    q = Polycos()
+    q.read_polyco_file(output_polyco)
+
+    mjds = np.linspace(55000.0, 55002.0, 11)
```
[the remark this answers is not in the document]
Comment (theXYZT) [reply tests/test_polycos.py:70]: I believe this throws an error. Let me add an explicit test for it.

File: tests/test_polycos.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    p.read_polyco_file(polyco_file)
+    p.write_polyco_file(output_polyco)
+
+    with open(output_polyco, "r") as f:
+        p2 = f.read()
+
+    assert p1 == p2
+
+
+def test_generate_polycos(tmpdir, par_file):
```
[the remark this answers is not in the document]
Comment (theXYZT) [reply tests/test_polycos.py:59]: Personally, I don't think that adds much beyond the tests that already exist. All this module does is read/write in the polyco format - which is mostly tested by the couple of round-trip tests already in there.

That being said, I am not really well-versed in using `hypothesis` and am still learning how to work with `PINT` in general - so someone else might have to add those tests if needed.

**Review by theXYZT — COMMENTED** [review]
Hopefully, I address all your comments @aarchiba !

File: tests/test_polycos.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aarchiba) [reply tests/test_polycos.py:72]: I'm less fussed about `hypothesis`, it seems overkill here, but I think it's valuable to check this works with a few non-default sets of parameters.

[the remark this answers is not in the document]
Comment (theXYZT) [reply tests/test_polycos.py:72]: I think I am misunderstanding what you mean. Can you provide an example of the kind of test you are suggesting here?

[the remark this answers is not in the document]
Comment (aarchiba) [reply tests/test_polycos.py:72]: Currently this test only generates polycos located at Arecibo, that are 300 minutes long, at 1400 MHz, et cetera. It's very possible that our polyco implementation breaks if asked to use, say, the barycenter, or longer or shorter segments, or... The easiest way to solve this is to use `@pytest.mark.parametrize` to run this test with several different combinations of parameters. Here is an example of doing this: https://github.com/nanograv/PINT/blob/4a0b6875483c5ceed1d8ea6228e535144a1f56fa/tests/test_pulse_number.py#L66

[the remark this answers is not in the document]
Comment (theXYZT) [reply tests/test_polycos.py:72]: Ah! I see. Sorry, I thought you were suggesting checking for correctness with these parameters. But if you just mean round-trip tests, then that should be easy.

[the remark this answers is not in the document]
Comment (aarchiba) [reply tests/test_polycos.py:72]: One could also check that the polyco prediction doesn't depend on the parameter settings, but since you're already checking against the model prediction that doesn't actually add anything.

(`hypothesis` could be used to thoroughly explore the space of possible combinations of inputs. But it's more complicated and you can spend a fair amount of time chasing non-problems.)

[the remark this answers is not in the document]
Comment (theXYZT) [reply tests/test_polycos.py:72]: Would you happen to know what obs_code tempo2 uses for barycenter/coe? I can't seem to figure this out.

[the remark this answers is not in the document]
Comment (paulray) [reply tests/test_polycos.py:72]: @ or bat

[the remark this answers is not in the document]
Comment (paulray) [reply tests/test_polycos.py:72]: For the geocenter I think it is 0 or coe

**Comment by theXYZT** [comment]
In talking to @mhvk, I learned that there's a small bug in my polyco implementation. So, I've converted this to a draft until fixed.

File: tests/test_polycos.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (theXYZT) [reply tests/test_polycos.py]: Done. Thanks for the comments!

**Comment by aarchiba** [comment]
Looks good to me. I think it's okay to merge once Travis eventually finishes.
